### PR TITLE
Correctly link URLs to users/repos with dashes, dots or underscores

### LINKED ISSUE #18888: URLs to issues/pull-requests for repos or users with dash in the name are not linked correctly
### Gitea Version

1.16.2

### Git Version

2.30.2

### Operating System

Debian 11

### How are you running Gitea?

I'm using the released version of Gitea downloaded from GitHub as a systemd service. The same issue reproduces on https://try.gitea.io

### Database

SQLite

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

-

### Description

Full URLs to issues and pull requests are not correctly linked and there is no hover preview when the owner's name or the repository name contains a dash. For example, create the repos `firstrepo` and `second-repo` on the demo instance, create an issue in each of them and then submit the following text as a new issue:
```
 * https://try.gitea.io/username/firstrepo/issues/1
 * https://try.gitea.io/username/second-repo/issues/1
```

It is rendered as:
```
 * username/firstrepo#1
 * https://try.gitea.io/username/second-repo/issues/1
```

### Screenshots

screenshot1
screenshot2

### PR DESCRIPTION
**Pull request by rtpt-alexanderneumann**
This PR extends the regex which is used to match full URLs in all texts to issues/PRs in repos hosted on Gitea so that user/repo names with dashes, dots or underscores (e.g. `http://localhost:5000/some-user/reponame`) are matched correctly. The first commit introduces tests that fail, the second commit extends the regex from `\w+` to `[\w_.-]+`.

I haven't found a central place where the user/repo name validation is performed. If there is a better way to do this please let me know. Thanks!

Closes #18888

### CODE REVIEW

File: modules/markup/html.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ var issueFullPatternOnce sync.Once
 func getIssueFullPattern() *regexp.Regexp {
 	issueFullPatternOnce.Do(func() {
 		issueFullPattern = regexp.MustCompile(regexp.QuoteMeta(setting.AppURL) +
-			`\w+/\w+/(?:issues|pulls)/((?:\w{1,10}-)?[1-9][0-9]*)([\?|#](\S+)?)?\b`)
+			`[\w_.-]+/[\w_.-]+/(?:issues|pulls)/((?:\w{1,10}-)?[1-9][0-9]*)([\?|#](\S+)?)?\b`)
```
Comment (KN4CK3R) [inline modules/markup/html.go:102]:
```suggestion
			`[\w_\.-]+/[\w_\.-]+/(?:issues|pulls)/((?:\w{1,10}-)?[1-9][0-9]*)([\?|#](\S+)?)?\b`)
```
`.` matches every character.

Comment (rtpt-alexanderneumann) [reply modules/markup/html.go:102]: Oh, good point (pun intended), thanks!

**Comment by KN4CK3R** [comment]
Sorry, mistake on my side. Inside `[ ]` the dot is simply a dot and does not need to be escaped.

**Comment by rtpt-alexanderneumann** [comment]
Ah, no worries, my test case was wrong as well :stuck_out_tongue_closed_eyes: 

I'll revert it.

**Comment by rtpt-alexanderneumann** [comment]
For the reference, here's my (now correct) simple test: https://go.dev/play/p/b4owdEOFUt3

**Review by noerw — APPROVED** [review]

File: modules/markup/html.go (hunk 1 above)
Comment (silverwind) [reply modules/markup/html.go:102]: This is incorrect, inside character classes, `.` (and all other special characters) have no special meaning, so this this qualifies as a useless escaping.

If you attempt the same regex in JS, the linter would [not like it](https://eslint.org/docs/rules/no-useless-escape), so I would revert this change.

Comment (silverwind) [reply modules/markup/html.go:102]: Ah, I see it was fixed already.

**Review by silverwind — APPROVED** [review]

**Comment by 6543** [comment]
:robot:

**Comment by lunny** [comment]
Please send backport

**Comment by silverwind** [comment]
Backport in https://github.com/go-gitea/gitea/pull/18908.

**Comment by Gusted** [comment]
Why is this marked to skip changelog?

**Comment by lunny** [comment]
@Gusted The fix will be in v1.16.3's changelog, so it should not be in 1.17's changelog.

**Pull request merged**